### PR TITLE
 Determine mount points without relying on /proc/mounts

### LINKED ISSUE #105642: Volume manager re-conciler takes too long to reconcile under high pressure and large number of subpaths
### What happened?

We have a situation where we are launching a lot of short-lived pods on a single node on EKS1.20 as a cronjob. Some volumes can take upto 2 hours to reconcile and unmount. Each pod has 3 volumes and 100 subPath

1. EFS volume
2. emptyDir 
3.  config map

Below logs show this
```
Oct 12 17:59:20 ip-192-168-58-97.ec2.internal kubelet[17955]: I1012 17:59:20.422730   17955 reconciler.go:193] Starting operationExecutor.UnmountVolume for volume "canikicka" [manu](UniqueName: "kubernetes.io/configmap/184df45d-548d-4012-8906-004779c9b22a-canikicka") pod "184df45d-548d-4012-8906-004779c9b22a" (UID: "184df45d-548d-4012-8906-004779c9b22a")
Oct 12 20:49:28 ip-192-168-58-97.ec2.internal kubelet[17955]: E1012 20:49:28.463764   17955 nestedpendingoperations.go:177] [manu] backOffErr: Failed to create operation with name "{volumeName:kubernetes.io/configmap/184df45d-548d-4012-8906-004779c9b22a-canikicka podName:184df45d-548d-4012-8906-004779c9b22a nodeName:}". An operation with that name failed at 2021-10-12 20:47:41.98837091 +0000 UTC m=+12608.434290319. No retries permitted until 2021-10-12 20:49:43.98837091 +0000 UTC m=+12730.434290319 (2m2s). Last error: "error cleaning subPath mounts for volume \"canikicka\" [manu](UniqueName: \"kubernetes.io/configmap/184df45d-548d-4012-8906-004779c9b22a-canikicka\") pod \"184df45d-548d-4012-8906-004779c9b22a\" (UID: \"184df45d-548d-4012-8906-004779c9b22a\") : error processing /var/lib/kubelet/pods/184df45d-548d-4012-8906-004779c9b22a/volume-subpaths/canikicka/init2: error cleaning subpath mount /var/lib/kubelet/pods/184df45d-548d-4012-8906-004779c9b22a/volume-subpaths/canikicka/init2/4: could not get consistent content of /proc/mounts after 3 attempts".
```

The above volume takes around 2 hours before it stops getting an unmount volume error.

### What did you expect to happen?

I was not expecting the reconciler to take 2 hours.

### How can we reproduce it (as minimally and precisely as possible)?

Same as https://github.com/kubernetes/kubernetes/issues/104976

### Anything else we need to know?

1. I tried the same with https://github.com/kubernetes/kubernetes/commit/82cfe9f14f8fb445d682ce2774ea44ce54885e81 ; however it is still not sufficient in some cases.

2. Since Unmount operations can happen concurrently on the code, they might change /proc/mounts very fast which which can throw back the errors to the volumemanager reconciler really fast while reading from /proc/mounts. is it possible to optimize subPath volumes during where a specif check for a given subpath; if a mount point is consistently present (2 consecutive reads) match for a mount point.
For eg
```
/proc/mounts has

Read 1
M1
M2
M3

Read 2
M2
M3 
M4
```

A consistentRead for M2, M3 will succeed while consistentRead for M1 or M4 will fail on the Unmount subPath volume path.

Thanks

cc @jsafrane @jingxu97 @thockin 

/sig storage

### Kubernetes version

<details>

1.20
</details>

### Cloud provider

<details>
AWS
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

[ec2-user@ip-192-168-58-97 ~]$ cat /etc/os-release
NAME="Amazon Linux"
VERSION="2"
ID="amzn"
ID_LIKE="centos rhel fedora"
VERSION_ID="2"
PRETTY_NAME="Amazon Linux 2"
ANSI_COLOR="0;33"
CPE_NAME="cpe:2.3:o:amazon:amazon_linux:2"
HOME_URL="https://amazonlinux.com/"
[ec2-user@ip-192-168-58-97 ~]$

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by manugupt1**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

use openat2 which is present in linux kernel 5.6+ to avoid
relying on /proc/mounts which is a slower operation.

For all other cases, fall back to rely on the list api which
consistently reads /proc/mounts for linux.

On a kernel 5.10, I can launch

With this case, I have been able to get around 5x more pod launches on
a single node. Earlier, I would be able to get ~400 pod launches with
performance degradation later ; but now; I have been able to launch >
2000 pods without any failure.

I was also able to launch pods much faster with these changes:
Earlier I was able to launch 400 pods in an hour
Now, I am able to launch 2000 pods in around 60 mins on Linux

#### Which issue(s) this PR fixes:

Fixes #105642

#### Special notes for your reviewer:

Follow up on sig-storage meeting.

#### Does this PR introduce a user-facing change?

```release-note
Linux kernel 5.6+ should experience faster pod launch and unlaunch times.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

/sig storage

/assign @jsafrane @jingxu97

### CODE REVIEW

**Review by thockin — COMMENTED** [review]
This is  a cute trick - how are we  going to test it?

**Review by mkimuram — COMMENTED** [review]
@manugupt1

Thank you for updating the code. Some additional comments.

File: staging/src/k8s.io/mount-utils/mount_linux_test.go [hunk 1]
```diff
@@ -545,3 +549,220 @@ func mountArgsContainOption(t *testing.T, mountArgs []string, option string) boo
...
+
+				symlink := filepath.Join(tmpDir, "symlink")
+				err = os.Symlink(f.Name(), symlink)
+				if err != nil {
+					return "", err
+				}
+				return symlink, nil
+			},
+		},
+		"path is a mount point": {
```
Comment (jingxu97) [inline staging/src/k8s.io/mount-utils/mount_linux_test.go:594]: are you able to test a bind mount case? Since it requires root, manual test is fine.

Also a bind mount to a symlink etc?

**Comment by kolyshkin** [comment]
I made a change in https://pkg.go.dev/github.com/moby/sys/mountinfo package to return ENOENT in case the path argument is not found, and incorporate other fixes discussed in here (see https://github.com/moby/sys/pull/90). Will do a tagged release with these changes soon.

It would be nice to vendor and use it here instead of forking the code. As you can see, the code is not that simple, and I'd rather have a single place to maintain it.

File: staging/src/k8s.io/mount-utils/mount_linux_test.go [hunk 2]
```diff
@@ -545,3 +549,220 @@ func mountArgsContainOption(t *testing.T, mountArgs []string, option string) boo
...
 }
+
+func TestIsMountPointUsingOpenAt2(t *testing.T) {
+	tests := map[string]struct {
+		expectedErr        error
+		expectedMountPoint bool
+		// returns the path to be tested for.
+		// returns an error if something went wrong during prepare.
+		prepare     func(*testing.T) (string, error)
+		openat2Mock func(int, string, *unix.OpenHow) (fd int, err error)
```
Comment (jaypipes) [inline staging/src/k8s.io/mount-utils/mount_linux_test.go:560]: @manugupt1 you can pull this mock out of here since it is unused.

**Comment by manugupt1** [comment]
> I made a change in https://pkg.go.dev/github.com/moby/sys/mountinfo package to return ENOENT in case the path argument is not found, and incorporate other fixes discussed in here (see [moby/sys#90](https://github.com/moby/sys/pull/90)). Will do a tagged release with these changes soon.
> 
> It would be nice to vendor and use it here instead of forking the code. As you can see, the code is not that simple, and I'd rather have a single place to maintain it.

Hi Kir, 

I will really like to have a single place to maintain the code . The issue is that moby/sys does not do a consistent check which I k8s really wants (https://github.com/moby/sys/blob/03b9f8d59a07f5206a2264105f4903a222aea964/mountinfo/mountinfo_linux.go#L130-L138);  I discussed not having a consistent check at sig-storage meeting and I was told it was not possible and there are several places online where we can find the discussion. I am not sure how to go about it while also using moby/sys. 

From what I see, there are 3 options:
1. k8s does maintain the code itself which is what this PR does.
2. moby can possibly export the fast path which will guarantee if a file is mounted via openat2 but if it cannot determine, k8s can fallback to its own consistency check.
3. moby/sys can implement the /proc/mounts consistency check and k8s can directly consume the mounted path.

cc @jsafrane @jingxu97

File: staging/src/k8s.io/mount-utils/mount_linux_test.go [hunk 3]
```diff
@@ -545,3 +549,220 @@ func mountArgsContainOption(t *testing.T, mountArgs []string, option string) boo
...
+			},
+		},
+	}
+
+	hasOpenAt2Support := isOpenAt2Supported()
+	if !hasOpenAt2Support {
+		t.Logf("openat2 is not supported, will use mock function openAt2Mock")
+	} else {
+		t.Log("openat2 is supported, will use openat2 syscall.")
+	}
```
Comment (jaypipes) [inline staging/src/k8s.io/mount-utils/mount_linux_test.go:654]: instead of the conditional below on lines 664-668, you can define the openat2 func here:

```go
type openat2fndef func(int, string, *unix.OpenHow) (fd int, err error)
var openat2fn openat2fndef = nil
if !isOpenAt2Supported() {
    openat2fn = openAt2Mock
}
```

File: staging/src/k8s.io/mount-utils/mount_linux_test.go [hunk 4]
```diff
@@ -545,3 +549,220 @@ func mountArgsContainOption(t *testing.T, mountArgs []string, option string) boo
...
+			continue
+		}
+
+		// let's not use mock if openat2 is supported on a CI.
+		var isMountPoint bool
+		if hasOpenAt2Support {
+			isMountPoint, err = isMountPointUsingOpenAt2(path, nil)
+		} else {
+			isMountPoint, err = isMountPointUsingOpenAt2(path, openAt2Mock)
+		}
```
Comment (jaypipes) [inline staging/src/k8s.io/mount-utils/mount_linux_test.go:668]: and then here, you only need to do:

```go
isMountPoint, err := isMountPointUsingOpenAt2(path, openat2fn)
```

File: staging/src/k8s.io/mount-utils/mount_linux_test.go [hunk 5]
```diff
@@ -545,3 +549,220 @@ func mountArgsContainOption(t *testing.T, mountArgs []string, option string) boo
...
+					test.expectedErr, err)
+			}
+		} else if err != nil {
+			t.Errorf("test (%s) did not expect any error to be returned but got (%s)", name,
+				err)
+		}
+	}
+}
+
+func openAt2Mock(dirfd int, path string, how *unix.OpenHow) (int, error) {
```
Comment (jaypipes) [inline staging/src/k8s.io/mount-utils/mount_linux_test.go:687]: I question the utility of using dependency injection for testing of a syscall, but ok... :)

**Comment by kolyshkin** [comment]
> I discussed not having a consistent check at sig-storage meeting and I was told it was not possible and there are several places online where we can find the discussion.

Can you please share a link (or a few)? I am very interested in the topic but don't remember reading anything this hack (re-reading mountinfo twice to ensure its consistency).

I guess moby can either export the fast path only, or add a call which re-reads mountinfo.

**Comment by manugupt1** [comment]
> > I discussed not having a consistent check at sig-storage meeting and I was told it was not possible and there are several places online where we can find the discussion.
> 
This is where this was added:
https://github.com/kubernetes/kubernetes/pull/3180/files

This is spoken here: https://github.com/kubernetes/kubernetes/issues/103754
A lot of conversation happened here as well: 
https://github.com/kubernetes/kubernetes/pull/105756

I believe the consistency check was introduced to mitigate:
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=593516

> Can you please share a link (or a few)? I am very interested in the topic but don't remember reading anything this hack (re-reading mountinfo twice to ensure its consistency).
> 
> I guess moby can either export the fast path only, or add a call which re-reads mountinfo.

That would be awesome; I think k8s can consume it if that is possible.

File: staging/src/k8s.io/mount-utils/mount_linux_test.go (hunk 1 above)
Comment (manugupt1) [reply staging/src/k8s.io/mount-utils/mount_linux_test.go:595]: I added the respective tests here as well so that you can pull and test to run it manually if reqd.

**Comment by jingxu97** [comment]
@kolyshkin do you have any comments  about consistent check? Whether it is possible to achieve it in moby?

**Comment by kolyshkin** [comment]
> @kolyshkin do you have any comments about consistent check? Whether it is possible to achieve it in moby?

Good question. Personally, I find re-reading mountinfo twice by default a bad idea, for multiple reasons:
 - it introduces a race (if someone starts/stops a container in between the two reads, the results from those will be different);
 - it creates a high load on the system (reading mountinfo is slow, reading it twice is twice as slow).

Having said all that, I agree that in some specific scenarios such re-reading can be helpful and might even be required. For `IsNotMountPoint`, the only such scenario is
 - the mount in question is a bind mount;
 - openat2 is not available;
 - kernel has a mountinfo bug (fixed in kernel 5.8, see https://github.com/kolyshkin/procfs-test for some more details and a repro).

This is because it's not possible to reliably determine whether the path is a mount point for a bind mount. To be more specific, reading mountinfo only once might lead to an occasional false negative. In this scenario, re-reading mountinfo again improves the chances of not getting a false positive, but it's still doesn't give us 100% reliability in all cases.

In addition to the above, re-reading mountinfo can be required when the caller is absolutely sure the path is mounted, but it can not be found in mountinfo. NOTE this is a _conditional_ re-read.

Currently, such re-read is _always_ performed which is ... suboptimal.

So, the action item here is to review k8s.io/mount-utils and its users, eliminate the double read for cases in which it is not needed (potentially changing its API in the process).

I'll try to take a look.

**Comment by kolyshkin** [comment]
> So, the action item here is to review k8s.io/mount-utils and its users, eliminate the double read for cases in which it is not needed (potentially changing its API in the process).

If we narrow this down to only determining mount points, what can be done is:

1. figure out which users of `IsNotMountPoint` need potential re-reading of mountinfo; the criteria are:
   - a user should care about bind mounts;
   - getting a false negatives will seriously disrupt workflow;
2. if any such users are found, implement something in moby/sys/mountinfo to support re-reading (most probably a separate function).
3. merge `IsNotMountPoint` and `IsLikelyNotMountPoint` into a single function, using moby/sys/mountinfo;

I can do 2 and 3, but not really sure about 1 -- need some help with it.

**Comment by kolyshkin** [comment]
> implement something in moby/sys/mountinfo to support re-reading (most probably a separate function).

PTAL https://github.com/moby/sys/pull/95

Item 1 above still needs to be worked on.

**Comment by mkimuram** [comment]
>> @kolyshkin do you have any comments about consistent check? Whether it is possible to achieve it in moby?

> Good question. Personally, I find re-reading mountinfo twice by default a bad idea, for multiple reasons:

An alternative way would be to make `mountedByOpenat2` global scope in moby/sys so that consumers, like k8s, can fall back to their own logics on errors. I mean if moby/sys provides `isMountFastPath` equivalent function, this PR would be able to change to just consume it in `IsNotMountPoint` in the same way.

I don't mean to disagree  to both of the existing approaches, or  (1) k8s implements its own entire logic including openat2 and (2) moby/sys implements fall back logic and k8s utilizes it. However, I just would like to share an alternative.

WDYT? @kolyshkin @jingxu97

File: staging/src/k8s.io/mount-utils/mount_linux.go [hunk 6]
```diff
@@ -667,3 +674,82 @@ func forceUmount(path string) error {
...
+	if openat2 == nil {
+		openat2 = unix.Openat2
+	}
+	// explicitly clean the path to remove trailing slashes.
+	// ref: https://pkg.go.dev/path/filepath#Clean
+	// otherwise, this will remove false +-ves or return errors.
+	// Ref: https://github.com/kubernetes/kubernetes/pull/105833#issuecomment-951092014
+	path = filepath.Clean(path)
+
+	// only work on absolute paths i.e starting with /
```
Comment (thockin) [inline staging/src/k8s.io/mount-utils/mount_linux.go:706]: I don't understand WHY this is important.  If it really is, can you comment on it here?

File: staging/src/k8s.io/mount-utils/mount_linux.go [hunk 7]
```diff
@@ -667,3 +674,82 @@ func forceUmount(path string) error {
...
+		// don't add the path to the error as it can be sensitive.
+		return false, errPathNotAbsolute
+	}
+
+	dir := filepath.Dir(path)
+	base := filepath.Base(path)
+
+	dirfd, err := openat2(-1, dir, &unix.OpenHow{
+		// O_PATH: indicates not to open the file and run operations solely on FD.
+		// O_CLOEXEC: will stop creating another file descriptor.
```
Comment (thockin) [inline staging/src/k8s.io/mount-utils/mount_linux.go:717]: This comment is not really accurate.  O_CLOEXEC means "close this FD on calls to exec()"

File: staging/src/k8s.io/mount-utils/mount_linux.go [hunk 8]
```diff
@@ -667,3 +674,82 @@ func forceUmount(path string) error {
...
+		Resolve: unix.RESOLVE_NO_SYMLINKS | unix.RESOLVE_NO_MAGICLINKS,
+	})
+	if err != nil {
+		return false, &os.SyscallError{Syscall: "openat2", Err: err}
+	}
+	defer unix.Close(dirfd)
+
+	fd, err := openat2(dirfd, base, &unix.OpenHow{
+		// O_PATH: indicates not to open the file and run operations solely on FD.
+		// O_CLOEXEC: will stop creating another file descriptor.
```
Comment (thockin) [inline staging/src/k8s.io/mount-utils/mount_linux.go:730]: same comment as above

**Comment by jingxu97** [comment]
> > So, the action item here is to review k8s.io/mount-utils and its users, eliminate the double read for cases in which it is not needed (potentially changing its API in the process).
> 
> If we narrow this down to only determining mount points, what can be done is:
> 
> 1. figure out which users of `IsNotMountPoint` need potential re-reading of mountinfo; the criteria are:
>    
>    * a user should care about bind mounts;
>    * getting a false negatives will seriously disrupt workflow;
> 2. if any such users are found, implement something in moby/sys/mountinfo to support re-reading (most probably a separate function).
> 3. merge `IsNotMountPoint` and `IsLikelyNotMountPoint` into a single function, using moby/sys/mountinfo;
> 
> I can do 2 and 3, but not really sure about 1 -- need some help with it.

@kolyshkin thanks for helping on this. From you analysis and  your proposed PR  moby/sys#95,  after kernel 5.8, re-reading mountinfo is not needed. So is it possible to conditionally trigger re-reading (if kernel is before 5.8)?

File: staging/src/k8s.io/mount-utils/mount_linux.go (hunk 6 above)
Comment (manugupt1) [reply staging/src/k8s.io/mount-utils/mount_linux.go:706]: This is important because we can cross different mount points and get either a false +ve or -ve if the path does not start with `/`;  I will add a comment as well with appropriate example.

**Comment by kolyshkin** [comment]
> @kolyshkin thanks for helping on this. From you analysis and your proposed PR [moby/sys#95](https://github.com/moby/sys/pull/95), after kernel 5.8, re-reading mountinfo is not needed. So is it possible to conditionally trigger re-reading (if kernel is before 5.8)?

If the only way was to use mountinfo, yes, it could make sense to add a kernel version check (although it's not always working because vendors tend to do backports etc).

Combined with openat2 method of detecting mount points, though, it is not needed, since openat2 is available since kernel 5.6. So, on kernels 5.6+ the fallback of parsing mountinfo won't be used anyway (unless in some corner case scenarios, say, if openat2 is denied via seccomp).

To sum it up, since kernel version detection is unreliable, and openat2 is introduced in an earlier kernel, such condition is not practically needed, it would just further complicate the code.

**Comment by manugupt1** [comment]
@jingxu97 @kolyshkin I think it is better to use openat2 than conditionally not re-read as both the changes are kernel related. I think we should objectively try to get openat2 change either exported in moby or through this PR. It would keep the code simpler and also keep the system performance a lot better.

File: staging/src/k8s.io/mount-utils/mount_linux.go (hunk 6 above)
Comment (jsafrane) [reply staging/src/k8s.io/mount-utils/mount_linux.go:706]: The comment still does not make much sense. It should say that it needs absolute path, I can read `if` below, it should says why it's important to have absolute paths.

File: staging/src/k8s.io/mount-utils/mount_linux.go (hunk 7 above)
Comment (jsafrane) [reply staging/src/k8s.io/mount-utils/mount_linux.go:717]: +1, O_CLOEXEC does not have sense in this context.

File: staging/src/k8s.io/mount-utils/mount_linux.go [hunk 9]
```diff
@@ -667,3 +674,82 @@ func forceUmount(path string) error {
...
+	return isMountPointUsingOpenAt2(path, nil)
+}
+
+// isMountPointUsingOpenAt2 uses openat2 syscall is a method of detecting a mount that works for all kinds
+// of mounts (incl. bind mounts), but requires a recent (v5.6+) linux kernel.
+// The code takes inspiration from here:
+// https://github.com/moby/sys/blob/40883be4345cf291513e2a4112896d393455fdfb/mountinfo/mounted_linux.go#L10
+// It expects that all symlinks have been evaluated.
+// If an ENOENT is returned by openat2, return os.ErrNotExist to satisfy isNotMountPoint return values.
+// openat2 is injected for testing purposes. In production code path, it should always be nil.
```
Comment (jsafrane) [inline staging/src/k8s.io/mount-utils/mount_linux.go:694]: Errors from `isMountFastPath` are always only compared to `nil` and then thrown away. They're not returned from `isNotMountPoint()`.

File: staging/src/k8s.io/mount-utils/mount_linux_test.go [hunk 10]
```diff
@@ -545,3 +550,293 @@ func mountArgsContainOption(t *testing.T, mountArgs []string, option string) boo
...
+				mountPath, err := os.MkdirTemp("", "target")
+				if err != nil {
+
+					return "", err
+				}
+				t.Cleanup(func() {
+					os.RemoveAll(mountPath)
+				})
+				mounter := New("")
+				err = mounter.Mount(sourceDir, mountPath, "", []string{"bind"})
```
Comment (jsafrane) [inline staging/src/k8s.io/mount-utils/mount_linux_test.go:669]: This requires the tests running as root. We avoid these. It means you can't really test with real `openat2()`, it must be always mock. I suggest adding a func with mock implementation into every test in this array instead of separate `openAt2Mock` with heuristics. And since all `openat2()` calls are simulated, IMO the tests don't need to touch anything on the filesystem.

(The hard part is getting all the errors from a real `openat2`, but it seems you have that part covered very well).

Comment (manugupt1) [reply staging/src/k8s.io/mount-utils/mount_linux_test.go:669]: Ah okay! Would you be able to review: https://github.com/moby/sys/pull/97 and very well maintained. From the conversations in this PR and elsewhere I think that is more desirable. 

It also simplifies a lot of the code. thanks for being patient.

**Comment by dims** [comment]
@manugupt1 looks like the plan is to get https://github.com/moby/sys/pull/97/files landed and then use `moby/sys` in k/k? (sounds good!)

**Comment by dims** [comment]
This PR needs a CLA to make progress, please see comments above from the CLA bot(s)

**Comment by dims** [comment]
There is a cncf-cla:no label on this PR. Can you please see the bot comments above to fix the CLA? This PR cannot progress unless the bot is happy with the CLA situation

**Comment by dims** [comment]
Please see https://github.com/kubernetes/community/blob/master/CLA.md#the-contributor-license-agreement on how to sign the CLA.

If you have already signed, please rebase commits in this PR to latest master and reopen the PR. This will trigger the newer EasyCLA bot to run and clear the "cncf-cla: no" label. We cannot merge any PRs if the CLA is not signed.

Thanks!
